feat: Support for AMP HTML Email

modernwarfareuplink opened the pull request:
I acknowledge that my contributions will be made under the same open-source license that the open-source project is provided under

# Adds support for sending AMP HTML email  #

- Added third mime type -> `text/x-amp-html`
- Send AMP HTML email just like normal html/plain email

### Resolves
- Fixes #940 
- Fixes #850 

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guidelines](https://github.com/sendgrid/sendgrid-python/blob/main/CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the main branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added the necessary documentation about the functionality in the appropriate .md file
- [x] I have added inline documentation to the code I modified

modernwarfareuplink commented:
@thinkingserious need your superpowers here !!! . Please check and let me know about the PR. Thanks

modernwarfareuplink commented:
@thinkingserious gentle reminder. Please let me know if you need any help. Thanks

thinkingserious commented on use_cases/sending_amp_html_content.py:
```diff
@@ -0,0 +1,99 @@
+'''
+
+message = Mail(
+    from_email='arun@raneesarees.in',
```
please change these email addresses to be generic, for example example@example.com.

thinkingserious commented on use_cases/sending_amp_html_content.py:
```diff
@@ -0,0 +1,99 @@
+    from_email='arun@raneesarees.in',
+    to_emails='aruntvs227@gmail.com',
+    subject='Sending with Twilio SendGrid is Fun',
+    html_content='<strong>and easy to do anywhere, even with Python</strong>')
```
I think it should be `html_content=amp_html_content`

modernwarfareuplink replied:
Thanks for helping out, will fix this up

thinkingserious commented on use_cases/sending_amp_html_content.py:
```diff
@@ -0,0 +1,99 @@
+# Sending AMP-HTML Email
+
+
+Currently, we require AMP HTML and any one of HTML or Plain Text content for improved deliverability or fallback for AMP HTML Email for supporting older clients and showing alternate content after 30 days. In some cases, only HTML may be available. The below example shows how to obtain the Plain Text equivalent of the HTML content.
```
This description does not match the code below. Maybe something like: "Following is an example on how to send an AMP HTML Email."

thinkingserious commented on sendgrid/helpers/mail/mail.py:
```diff
@@ -725,9 +730,16 @@ def add_content(self, content, mime_type=None):
         if isinstance(content, str):
             content = Content(mime_type, content)
-        # Content of mime type text/plain must always come first
+        # Content of mime type text/plain must always come first, followed by text/x-amp-html and then text/html
```
We should add a test to ensure this is true.

Also, I think `text/x-amp-html` should be last.

modernwarfareuplink replied:
Hmm, not really, if the `text/x-amp-html` is at the last, it fails in apple mail. Have tested the ordering earlier.

thinkingserious reviewed (changes requested):
Thank you!

modernwarfareuplink commented:
Thanks for the review @thinkingserious. Will update in 2 days, caught up with something. Will let you know, once the changes are done.

thinkingserious commented:
That sounds good @modernwarfareuplink, thank you!

thinkingserious reviewed (commented):
LGTM! Thank you!

Could you please update the tests here to finish this one off? https://github.com/sendgrid/sendgrid-python/blob/HEAD/test/test_mail_helpers.py 

Thanks again!

modernwarfareuplink commented:
Thanks for the review. I am really sorry that it took this long and no updates from my side. Will finish adding the tests by morning (in around 12 hours) and let you know.
Thank you and happy new year @thinkingserious

modernwarfareuplink commented:
@thinkingserious  have resolved your comments, made few changes around the ordering and added appropriate tests. Also found that the build was failing due to copyrights year and fixed it too. Please review and let me know if you need any additional details. Thank you

modernwarfareuplink commented:
@thinkingserious gentle reminder for review. Please let me know if you have any issues. Thank you

JenniferMah commented on test/test_mail_helpers.py:
```diff
@@ -284,6 +318,239 @@ def test_multiple_emails_to_multiple_recipients(self):
+        }
+        self.assertEqual(
+            message.get(),
+            json.loads(json.dumps(response_content))
```
For parity with the other tests in this file instead of using `json.dumps` to convert `response_content` use a raw string as seen in this [test](https://github.com/sendgrid/sendgrid-python/blob/HEAD/test/test_mail_helpers.py#L77).

JenniferMah commented on test/test_mail_helpers.py:
```diff
@@ -284,6 +318,239 @@ def test_multiple_emails_to_multiple_recipients(self):
+        }
+        self.assertEqual(
+            message.get(),
+            json.loads(json.dumps(response_content))
```
Please replace `json.dumps(response_content)` with the `response_content` as a raw string.

JenniferMah reviewed (changes requested):
Thanks for updating the tests! I've just requested a few changes but over all this looks good to me!

modernwarfareuplink commented:
Have updated the minor comment @JenniferMah . Please do have a look once. Thank you

modernwarfareuplink commented on test/test_mail_helpers.py:
```diff
@@ -284,6 +318,239 @@ def test_multiple_emails_to_multiple_recipients(self):
+                '<strong>and easy to do anywhere, even with Python</strong>')
+        )
+
+        response_content = json.dumps({
```
Have added it this way as the amp_html_content has other string delimiters and this simply looks cleaner and I think this is easy to update/maintain

modernwarfareuplink commented:
@thinkingserious need your superpowers for review and merging the changes 🙏

modernwarfareuplink commented:
Thank you @JenniferMah
